draft-how-does-mutant-work.markdown

```diff
@@ -0,0 +1,431 @@
+
+When it comes to mutation testing I learned one thing. It's complex. Really complex. And as a consequence, `Mutant` is complex. 
+I did learn a lot from reading it's source and I'd like to share this with you.
+Beware though, this is going to be a long ride - but at the end, it'll be worth your while.
+
+While doing research for my presentation, I wrote an introductory series of articles about the gems `Mutant` uses ([part 1](https://troessner.svbtle.com/lessons-learned-from-some-of-the-best-ruby-codebases-out-there-part-1), [part 2](https://troessner.svbtle.com/lessons-learned-from-some-of-the-best-ruby-codebases-out-there-part-2) and [part 3](https://troessner.svbtle.com/lessons-learned-from-some-of-the-best-ruby-codebases-out-there-part-3)). It probably makes sense if you check the series out first to be able to fully follow everything down below.
+
+Ok. Ready? Then let's roll.
+
+### Our entry point: bin/mutant
+
+The great thing about gems with a binary is that it's a lot easier to start exploring them - just start at the binary.
+
+Let's look at bin/mutant:
+
+```Ruby
+require 'mutant'
+
+namespace =
+  if ARGV.include?('--zombie')
+    $stderr.puts('Running mutant zombified!')
+    Mutant::Zombifier.call(
+      namespace:        :Zombie,
+      load_path:        $LOAD_PATH,
+      kernel:           Kernel,
+      pathname:         Pathname,
+      require_highjack: Mutant::RequireHighjack
+        .method(:call)
+        .to_proc
+        .curry
+        .call(Kernel),
+      root_require:     'mutant',
+      includes: %w[
+        mutant
+        unparser
+        morpher
+        adamantium
+        equalizer
+        anima
+        concord
+      ]
+    )
+    Zombie::Mutant
+  else
+    Mutant
+  end
+
+Kernel.exit(namespace::CLI.run(ARGV))
+```
+
+Ok, that looks friggin complicated already. Let's keep things simple and ignore everything that relates to the, uhm, "zombie". This would warrant a blog post of its own. 
+Here's our new executable:
+
+```Ruby
+require 'mutant'
+
+namespace =
+  if ARGV.include?('--zombie')
+    # Stuff we are going to ignore
+    # ...snip
+    Zombie::Mutant
+  else
+    Mutant
+  end
+
+Kernel.exit(namespace::CLI.run(ARGV))
+```
+
+That looks a lot easier to understand. So we're basically just using the regular `Mutant` namespace in case we didn't do the zombie thing so basically it boils down to the last line looking like this:
+
+
+```Ruby
+Kernel.exit(Mutant::CLI.run(ARGV))
+```
+
+which translates to:
+
+* Pass all of our CLI arguments to `Mutant::CLI.run`
+* `Mutant::CLI.run` runs the program and then returns the exit status
+* Which we pass to `Kernel#exit` that will cause our process to terminate with this very exit status
+
+Let's check out `Mutant::CLI.run`:
+
+```Ruby
+module Mutant
+  class CLI
+    include Adamantium::Flat, Equalizer.new(:config), Procto.call(:config)
+
+    def self.run(arguments)
```

Comment:
Actually it turns out `.run` is not misnamed. Since `.call` is taken and returns the parsed config.

`CLI` is actually a "CLI parser" object (and also documented as such), interface to invoke the parser is `.call(args)`.

`CLI.run` is the interface to both "parse and run" the `args`.

`CLI` likely should be split into `CLI::Parser` and `CLI::Runner` both with `.call` interface. But conceptually its not that bad as it initially sounded right now.